### PR TITLE
fix(tianmu)Question with date format of NULL (#682)

### LINKED ISSUE #682: bug: DELETE and SELECT query should have the same logic.
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

when SELECT .. where is NULL, it returns an empty set.
Something unexpected occurred, when DELETE .. where is NULL , it returns 1 row affected.

as follows
```shell
mysql> select * from t1;
+------------+
| a          |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)

mysql> select * from t1 where a is null;
Empty set (0.00 sec)

mysql> delete from t1 where a is null;
Query OK, 1 row affected (0.00 sec)

mysql> select * from t1;
Empty set (0.00 sec)
```

### Expected behavior

when SQL (SELECT .. where is NULL)  returns an empty set,
the data should not be deleted by the SQL(DELETE .. where is NULL)

```shell
mysql> select * from t1 where a is null;
+------------+
| a          |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)

mysql> delete from t1 where a is null;
Query OK, 1 row affected (0.01 sec)

```

### How To Reproduce

```sql
create table t1(a date not null);
insert ignore into t1 values (0);
select * from t1 where a is null;
delete from t1 where a is null;
```

### Environment

ubuntu@ubuntu:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"

ubuntu@ubuntu:~$ /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld Ver 5.7.36_v1.0.1_beta_2-StoneDB for Linux on x86_64 (build-)
build information as follow:
Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
Branch name: stonedb-5.7-dev
Last commit ID: https://github.com/stoneatom/stonedb/commit/208f79d80bae1667d147e0713950317c9ca6c6b9
Last commit time: Date: Sat Sep 17 16:19:42 2022 +0800
Build time: Date: Fri Sep 23 14:14:27 CST 2022

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by konghaiya**
Cause:
When the date in MySQL is 0, the value is considered to be null. MySQL handles this problem. If you query the statement of isull,
 you will convert isnull to=0. This problem is that the value with the date of 0 is inserted into the table.
 The query and deletion results of the tianmu engine are not consistent.
The reason for the inconsistency is that the syntax parsing logic of the query and deletion of tianmu is not the same

>Supplementary description of date type on MySQL official website:
lQLPJxbJU4noaETNBEfNCCWwjasZm-ZjXkkDSutaV4CqAA_2085_1095

Solution:
Unified query logic and modification logic for date format processing.
Keep the phenomenon of tianmu consistent with innodb.

## Summary about this PR

Issue Number: close #682

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Review by isredstar — APPROVED** [review]

File: mysql-test/suite/tianmu/t/issue682.test [hunk 1]
```diff
@@ -0,0 +1,57 @@
+--source include/have_tianmu.inc
+
+use test;
+set sql_mode='STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';
+create table t1(a date not null);
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue682.test:5]: How about the not `date`  data type ? such as `datetime` or `int` or `timestamp`, etc.

Comment (RingsC) [reply mysql-test/suite/tianmu/t/issue682.test:5]: >Invalid DATE, DATETIME, or TIMESTAMP values are converted to the “zero” value of the appropriate type ('0000-00-00' or '0000-00-00 00:00:00'), if the SQL mode permits this conversion. The precise behavior depends on which if any of strict SQL mode and the [NO_ZERO_DATE](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_date) SQL mode are enabled; see [Section 5.1.11, “Server SQL Modes”](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html).

How about this?

File: storage/tianmu/core/query.cpp [hunk 2]
```diff
@@ -1621,6 +1621,54 @@ int Query::BuildConditions(Item *conds, CondID &cond_id, CompiledQuery *cq, cons
   return RCBASE_QUERY_ROUTE;
 }
 
+bool Query::internalDataTimeIsNullToEq(THD *thd, Item *cond, Item **retcond) {
```
Comment (RingsC) [inline storage/tianmu/core/query.cpp]: i think that processing logic of datetime is beeter in `remove_eq_conds`.  and what the real value of `datetime` is stored in tianmu when we don't set the value and it allows null.

File: storage/tianmu/core/query_compile.cpp [hunk 3]
```diff
@@ -1078,6 +1078,7 @@ int Query::Compile(CompiledQuery *compiled_query, SELECT_LEX *selects_list, SELE
       if (!AddOrderByFields(order, tmp_table, group != nullptr || sl->join->select_distinct || aggr_used))
         throw CompilationError();
       CondID cond_id;
+      internalDataTimeIsNullToEq(sl->master_unit()->thd, conds, &conds);
```
Comment (RingsC) [inline storage/tianmu/core/query_compile.cpp]: if you want to do the type test, i think this test should be in `BuildConditions`.

File: mysql-test/suite/tianmu/t/issue682.test (hunk 1 above)
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue682.test:5]: and as the mysql docuemnt said, disable in `NO_ZERO_DATE` mode, we should do some tests for this mode 
and `NO_ZERO_DATE` mode.

**Review by RingsC — DISMISSED** [review]

File: storage/tianmu/core/query.cpp (hunk 2 above)
Comment (konghaiya) [reply storage/tianmu/core/query.cpp]: At present, our tianmu engine only optimizes the conditions in the join object, but does not optimize the conditions in (select_lx). BuildConditions directly uses the conditions in (select_lx). At present, this modification method has the least impact on the system

Comment (konghaiya) [reply storage/tianmu/core/query.cpp]: Use the condition in the join when the condition in the join is not null

File: mysql-test/suite/tianmu/t/issue682.test [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+-- error 1292
+insert into t1 values (0);
+insert ignore into t1 values (null);
+-- error 1048
+insert into t1 values (null);
+insert into t1 values (20221020);
+select * from t1;
+select * from t1 where a is null;
+-- error 1292
+delete from t1 where a is null;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue682.test:122]: Pls, add some complex condition statements.

**Review by RingsC — DISMISSED** [review]

File: mysql-test/suite/tianmu/t/issue682.test (hunk 4 above)
Comment (konghaiya) [reply mysql-test/suite/tianmu/t/issue682.test:122]: Added

**Review by RingsC — APPROVED** [review]
LGTM

**Pull request closed without merging**